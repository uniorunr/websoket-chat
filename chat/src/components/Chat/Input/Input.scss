.input-section {
  width: 100%;
  display: flex;
  justify-content: center;
  border: 2px solid black;
  border-radius: 0 0 5px 5px;
  font-size: 1.4rem;

  textarea {
    border: none;
    width: 100%;
    padding: 10px 20px;
    font-size: 1.4rem;
    border-radius: 0 0 5px 5px;
    background: #f1f1f1;
    outline: none;
    font-family: Roboto, "Helvetica Neue", "Apple Color Emoji", Helvetica, Arial, sans-serif;
    resize: none;
  }

  textarea:focus {
    background: #ffffff;
  }

  textarea:focus::placeholder {
    color: transparent;
  }

  textarea:focus:-moz-placeholder {
    color: transparent;
  }

  button {
    background: #000000;
    border: none;
    color: #ffffff;
    text-transform: uppercase;
    padding: 10px 20px;
    outline: none;
    cursor: pointer;
    border-left: 2px solid black;
  }

  button:hover {
    font-weight: 700;
    background: #ffffff;
    color: #000000;
    border-radius: 0 0 5px 0;
  }
}
