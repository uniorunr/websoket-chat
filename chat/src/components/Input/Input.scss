.input-section {
  width: 100%;
  display: flex;
  justify-content: center;
  border: 2px solid black;
  border-radius: 0 0 5px 5px;
  font-size: 1.4rem;

  input {
    border: none;
    width: 100%;
    padding: 10px 20px;
    font-size: 1.4rem;
    border-radius: 0 0 5px 5px;
    background: #f1f1f1;
    outline: none;
  }

  input:focus {
    background: #ffffff;
  }

  input:focus::-webkit-input-placeholder {
    color:transparent;
  }

  input:focus:-moz-placeholder {
    color:transparent;
  }

  input:focus::-moz-placeholder {
    color:transparent;
  }

  input:focus:-ms-input-placeholder {
    color:transparent;
  }

  button {
    background: #000000;
    border: none;
    color: #ffffff;
    text-transform: uppercase;
    padding: 10px 20px;
    outline: none;
    cursor: pointer;
    border-left: 2px solid black;
  }

  button:hover {
    font-weight: 700;
    background: #ffffff;
    color: #000000;
    border-radius: 0 0 5px 0;
  }
}
