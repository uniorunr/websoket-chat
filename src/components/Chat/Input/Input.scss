@import "../../../theme/colors";
$width_picker: 200px;

.input-section {
  position: relative;
  width: 100%;
  display: flex;
  justify-content: center;
  border: 2px solid $black-color;
  border-radius: 0 0 5px 5px;
  font-size: 1.4rem;
  background: $white-color;

  textarea {
    border: none;
    width: 100%;
    padding: 10px 20px 5px;
    font-size: 1.4rem;
    border-radius: 0 0 5px 5px;
    background: $input-color;
    outline: none;
    font-family: Roboto, "Helvetica Neue", "Apple Color Emoji", Helvetica, Arial, sans-serif;
    resize: none;
  }

  textarea:focus {
    background: $white-color;
  }

  textarea:focus::placeholder {
    color: transparent;
  }

  textarea:focus:-moz-placeholder {
    color: transparent;
  }

  .submit {
    background: $black-color;
    border: none;
    color: $white-color;
    text-transform: uppercase;
    padding: 10px 20px;
    outline: none;
    cursor: pointer;
    border-left: 2px solid $black-color;
  }

  .submit:hover {
    font-weight: 700;
    background: $white-color;
    color: $black-color;
    border-radius: 0 0 5px 0;
  }

  .picker-wrapper {
    visibility: hidden;

    .emoji-picker {
      position: absolute;
      width: $width_picker;
      bottom: 55px;
      right: 40px;

      .emoji-category {
        width: $width_picker;
      }
    }
  }

  .picker-wrapper.visible {
    visibility: visible;
  }

  .emoji-picker-icon {
    background: $input-color;
    border: none;
    text-transform: uppercase;
    padding: 0 20px 0 15px;
    outline: none;
    cursor: pointer;
  }
}

@media only screen and (max-width : 549px) {
  .input-section {
    .emoji-picker-icon {
      padding: 0 10px 0 5px;
    }
  }
}
