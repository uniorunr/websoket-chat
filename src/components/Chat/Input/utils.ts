const insertAtCursor = (input: HTMLTextAreaElement, textToInsert: string) => {
  const value = input.value;
  const start: number | null = input.selectionStart;
  const end: number | null = input.selectionEnd;
  input.value = value.slice(0, start) + textToInsert + value.slice(end);
  input.selectionStart = input.selectionEnd = start + textToInsert.length;
  input.focus();
};

const sendMessage = (
  ws: WebSocket | null,
  userName: string,
  chatStatus: string,
  offlineHandler: Function,
  textarea: HTMLTextAreaElement | null
): void => {
  if (
    ws &&
    ws.readyState === ws.OPEN &&
    textarea &&
    textarea.value &&
    chatStatus === 'online'
  ) {
    ws.send(
      JSON.stringify({
        from: userName,
        message: textarea.value
      })
    );
    if (textarea) {
      textarea.value = '';
    }
  } else if (chatStatus === 'offline' && textarea && textarea.value) {
    offlineHandler({
      from: userName,
      message: textarea.value
    });
    if (textarea) {
      textarea.value = '';
    }
  }
};

export { insertAtCursor, sendMessage };
