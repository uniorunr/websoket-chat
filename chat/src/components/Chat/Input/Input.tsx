import React, { Component } from 'react';
import { StoreState } from '../../../types';
import { connect } from 'react-redux';
import './Input.scss';

class Input extends Component<InputProps> {
  private inputRef: React.RefObject<HTMLInputElement> = React.createRef();

  componentDidMount(): void {
    document.addEventListener('keydown', this.keyboardListener);
  }

  componentWillUnmount(): void {
    document.removeEventListener('keydown', this.keyboardListener);
  }

  keyboardListener = (event: KeyboardEvent) => {
    if (
      event.code === 'Enter' &&
      document.activeElement === this.inputRef.current
    ) {
      this.sendMessage();
    }
  };

  sendMessage = () => {
    const { ws, userName } = this.props;
    if (
      ws &&
      ws.readyState === ws.OPEN &&
      this.inputRef.current &&
      this.inputRef.current.value
    ) {
      ws.send(
        JSON.stringify({
          from: userName,
          message: this.inputRef.current.value
        })
      );
      if (this.inputRef.current) {
        this.inputRef.current.value = '';
      }
    }
  };

  render() {
    return (
      <div className="input-section">
        <input placeholder="Type something..." ref={this.inputRef} />
        <button type="button" onClick={this.sendMessage}>
          send
        </button>
      </div>
    );
  }
}

interface InputProps {
  ws: WebSocket | null;
  userName: string;
}

const mapStateToProps = (state: StoreState) => ({
  ws: state.webSocketInstance,
  userName: state.userName
});

export default connect(mapStateToProps)(Input);
