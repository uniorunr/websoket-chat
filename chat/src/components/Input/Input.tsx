import React, { Component } from 'react';
import { StoreState } from '../../types';
import { connect } from 'react-redux';
import './Input.scss';

class Input extends Component<InputProps> {
  private inputRef: React.RefObject<HTMLInputElement> = React.createRef();

  componentDidMount(): void {
    document.addEventListener('keydown', this.keyboardListener);
  }

  componentWillUnmount(): void {
    document.removeEventListener('keydown', this.keyboardListener);
  }

  keyboardListener = (event: KeyboardEvent) => {
    if (
      event.code === 'Enter' &&
      document.activeElement === this.inputRef.current
    ) {
      this.sendMessage();
    }
  };

  sendMessage = () => {
    const { ws } = this.props;
    if (ws && ws.readyState === ws.OPEN) {
      ws.send(
        JSON.stringify({
          from: 'yourMom',
          message: this.inputRef.current ? this.inputRef.current.value : null
        })
      );
      if (this.inputRef.current) {
        this.inputRef.current.value = '';
      }
    }
  };

  render() {
    return (
      <div className="input-section">
        <input placeholder="Type something..." ref={this.inputRef} />
        <button type="button" onClick={this.sendMessage}>
          send
        </button>
      </div>
    );
  }
}

interface InputProps {
  ws: WebSocket | null;
}

const mapStateToProps = (state: StoreState) => ({
  ws: state.webSocketInstance
});

export default connect(mapStateToProps)(Input);
