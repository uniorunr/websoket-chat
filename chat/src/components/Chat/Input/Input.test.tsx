import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import store from '../../../store/mainStore';
import Input from './Input';

it('Input component renders correctly', () => {
  const app = shallow(
      <Provider store={store}>
        <Input />
      </Provider>
  );
  expect(app).toMatchSnapshot();
});
