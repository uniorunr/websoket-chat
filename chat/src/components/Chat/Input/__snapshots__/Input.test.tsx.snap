// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component renders correctly 1`] = `ShallowWrapper {}`;
